server/server.go

```diff
@@ -0,0 +1,214 @@
+	return func(s *Server) error {
+		s.initializeTimeout = timeout
+		return nil
+	}
+}
+
+// WithInitializer adds an initialization function that will get called prior to serving.
+func WithInitializer(initializerFunc InitializerFunc) Option {
+	return func(s *Server) error {
+		s.initializers = append(s.initializers, initializerFunc)
+		return nil
+	}
+}
+
+// WithGrpcServer adds the given GRPC server to this server. There can only be one GRPC server within a given instance,
+// so multiple calls with this option will overwrite the previous ones.
+func WithGrpcServer(grpcServer *grpc.Server) Option {
+	return func(s *Server) error {
+		s.GRPCServer = grpcServer
+		return nil
+	}
+}
+
+// WithHandler registers the given http handler to this server by registering the pattern at the root of the http server
+func WithHandler(pattern string, handler http.Handler) Option {
+	return func(s *Server) error {
+		s.registrars = append(s.registrars, func(mux *http.ServeMux) error {
+			mux.Handle(pattern, handler)
+			return nil
+		})
+		return nil
+	}
+}
+
+// WithHealthChecks registers the given health checker with this server by registering its endpoints at the root of the
+// http server.
+func WithHealthChecks(checker health.Checker) Option {
+	return func(s *Server) error {
+		s.registrars = append(s.registrars, func(mux *http.ServeMux) error {
+			checker.RegisterHandler(mux)
+			return nil
+		})
+		return nil
+	}
+}
+
+// WithGateway registers the given gateway options with this server
+func WithGateway(options ...gateway.Option) Option {
+	return func(s *Server) error {
+		s.registrars = append(s.registrars, func(mux *http.ServeMux) error {
+			_, err := gateway.NewGateway(append(options, gateway.WithMux(mux))...)
+			return err
+		})
+		return nil
+	}
+}
+
+// Serve invokes all initializers then serves on the given listeners.
+//
+// If a listener is left blank, then that particular part will not be served.
+//
+// If a listener is specified for a part that doesn't have a corresponding server, then an error will be returned. This
+// can happen, for instance, whenever a gRPC listener is provided but no gRPC server was set or no option was passed
+// into NewServer.
+func (s *Server) Serve(grpcL, httpL net.Listener) error {
+	if err := s.initialize(); err != nil {
+		return err
+	}
+	errC := make(chan error)
+
+	if httpL != nil {
+		if s.HTTPServer == nil {
+			return errors.New("httpL is specified, but no HTTPServer is provided")
+		}
+		go func() { errC <- s.HTTPServer.Serve(httpL) }()
+	} else {
+		s.HTTPServer = nil
+	}
+
+	if grpcL != nil {
+		if s.GRPCServer == nil {
+			return errors.New("grpcL is specified, but no GRPCServer is provided")
+		}
+		go func() { errC <- s.GRPCServer.Serve(grpcL) }()
+	} else {
+		s.GRPCServer = nil
+	}
+
+	return <-errC
+}
+
+func (s *Server) Stop() error {
+	doneC := make(chan bool)
+	errC := make(chan error)
+	go func() {
+		defer func() { doneC <- true }()
```

Comment:
Why do you actually need a deferred call here?

Reply:
you're right: I probably don't. I just blindly followed the concurrency patterns I found with wait groups. 

I kind of like it like this, though: in general it means that you can do whatever the hell you want below the line and still be sure that your `wg.Done()` gets called. I'm gonna keep it like this :)